File: examples/sentence-splitter/sentence_splitter.cpp
#include <cctype>
#include <iostream>
#include <iterator>
#include <set>
#include <vector>

void splitSentences(std::string input) {
    if (input.length() < 3) {
        // Just assume that it's a single sentence.
        std::cout << input << std::endl;
    }

    size_t pos = 0;                     // The period we are considering.
    size_t sentence_boundary = 0;       // The start of the current sentence.

    while ((pos = input.find_first_of(".?!", pos + 1)) != std::string::npos) {
        // Periods followed by whitespace followed by a lower case letter are
        // not sentence boundaries.
        if (pos + 2 < input.length() &&
                std::isspace(input.at(pos + 1)) &&
                std::islower(input.at(pos + 2))) {
            continue;
        }

        // Periods followed by a digit with no intervening whitespace are not
        // sentence boundaries.
        if (pos + 1 < input.length() &&
                std::isdigit(input.at(pos + 1))) {
            continue;
        }

        // Periods followed by whitespace and then an upper case letter, but
        // preceded by any of a short list of titles are not sentence
        // boundaries.
        const std::set<std::string> titles = {"Mr.", "Mrs.", "Dr.", "Jr."};
        if ((pos >= 2 && titles.count(input.substr(pos - 2, 3))) ||
                (pos >= 3 && titles.count(input.substr(pos - 3, 4)))) {
            continue;
        }

        // Periods internal to a sentence of letters with no adjacent
        // whitespace are not sentence boundaries (e.g., websites).
        if (pos + 1 < input.length() &&
                std::isalpha(input.at(pos - 1)) &&
                std::isalpha(input.at(pos + 1))) {
            continue;
        }

        // Periods followed by certain kinds of punctuation (notably comma and
        // more periods) are probably not sentence boundaries.
        if (input.at(pos + 1) == ',') {
            continue;
        }

        for (; pos + 1 < input.length() && input.at(pos + 1) == '.'; ++pos) {
            // Do nothing, we are just trying to consume any periods that are
            // next to each other.
        }

        // We passed all the special conditions, so we probably have a
        // sentence.
        std::string sentence =
            input.substr(sentence_boundary, pos - sentence_boundary + 1);
        std::cout << sentence << std::endl;

        // Start the next sentence after the period and the initial space.
        sentence_boundary = pos + 2;
    }
}

int main() {
    std::istream_iterator<char> end_of_stream;
    std::istream_iterator<char> input_it (std::cin >> std::noskipws);

    std::string input(input_it, end_of_stream);
    splitSentences(input);
}

File: examples/sentence-splitter/sentence_splitter.rs
#![feature(old_io,unicode)]

use std::old_io as io;
use std::collections::HashSet;

fn split_sentences(input: String) {
    if input.len() < 3 {
        // Assume that the input is a single sentence.
        println!("{}", input);
    }

    let chars: Vec<char> = input.chars().collect();
    let mut sentence_boundary = 0;

    // Find the next plausible sentence boundary.
    for (mut pos, _) in chars.iter().enumerate().filter(
            |&(_, &c)| c == '.' || c == '!' || c == '?') {
        // Periods followed by whitespace followed by a lower case letter are
        // not sentence boundaries.
        if pos + 2 < input.len() &&
              chars[pos + 1].is_whitespace() &&
              chars[pos + 2].is_lowercase() {
            continue
        }

        // Periods followed by a digit with no intervening whitespaces are not
        // sentence boundaries.
        if pos + 1 < input.len() && chars[pos + 1].is_digit(10) {
            continue
        }

        // Periods folled by whitespace and then an upper case letter, but
        // preceded by any of a short list of titles are not sentence
        // boundaries.
        let titles: HashSet<&'static str> =
                ["Mr.", "Mrs.", "Dr.", "Jr."].iter().map(|&x| x).collect();
        if pos >= 2 {
            let two_previous_chars: String =
                chars[pos-2..pos+1].iter().map(|&x| x).collect();
            if titles.contains(&two_previous_chars[..]) {
                continue
            }
        }
        if pos >= 3 {
            let three_previous_chars: String =
                chars[pos-3..pos+1].iter().map(|&x| x).collect();
            if titles.contains(&three_previous_chars[..]) {
                continue
            }
        }

        // Periods internal to a sentence of letter with no adjacent whitespace
        // are not sentence boundaries (e.g., websites).
        if pos + 1 < input.len() &&
              chars[pos + 1].is_alphabetic() &&
              chars[pos - 1].is_alphabetic() {
            continue
        }

        if pos + 1 < input.len() && chars[pos + 1] == ',' {
            continue
        }

        // Do nothing, we are just trying to consume any periods that are next
        // to each other.
        while pos + 1 < input.len() && chars[pos + 1] == '.' {
            pos += 1;
        }

        // We passed all the special conditions, so we probably have a sentence.
        if sentence_boundary < pos {
            let sentence: String =
                    chars[sentence_boundary..pos + 1].iter()
                    .map(|&x| x).collect();
            println!("{}", sentence);
        }

        sentence_boundary = pos + 2;
    }
}

fn main() {
    let mut stdin = io::stdin();
    let input = String::from_utf8(stdin.read_to_end().unwrap()).unwrap();
    split_sentences(input);
}
